src/NuGet.Core/NuGet.Build.Tasks.Pack/Pack.targets

```diff
@@ -56,24 +57,49 @@ Copyright (c) .NET Foundation. All rights reserved.
     </TfmSpecificPackageFile>
   </ItemDefinitionGroup>
 
+  <Target Name="_GetOutputItemsFromPack"
+          Returns="@(_OutputPackItems)">
+
+    <ConvertToAbsolutePath Paths="$(PackageOutputPath)">
+      <Output TaskParameter="AbsolutePaths" PropertyName="PackageOutputAbsolutePath" />
+    </ConvertToAbsolutePath>
+
+    <ConvertToAbsolutePath Paths="$(BaseIntermediateOutputPath)">
+      <Output TaskParameter="AbsolutePaths" PropertyName="NuspecOutputAbsolutePath" />
+    </ConvertToAbsolutePath>
+
+    <GetPackOutputItemsTask
+        PackageOutputPath="$(PackageOutputAbsolutePath)"
+        NuspecOutputPath="$(NuspecOutputAbsolutePath)"
+        PackageId="$(PackageId)"
+        PackageVersion="$(PackageVersion)"
+        IncludeSymbols="$(IncludeSymbols)"
+        IncludeSource="$(IncludeSource)">
+
+      <Output
+        TaskParameter="OutputPackItems"
+        ItemName="_OutputPackItems" />
+    </GetPackOutputItemsTask>
+  </Target>
+
   <!--
     ============================================================
     Pack
     Post Build Target
     ============================================================
   -->
   <Target Name="_PackAsBuildAfterTarget"
-          AfterTargets="Build" 
+          AfterTargets="Build"
           Condition="'$(GeneratePackageOnBuild)' == 'true' AND '$(IsInnerBuild)' != 'true'"
           DependsOnTargets="Pack">
   </Target>
 
   <Target Name="_CleanPackageFiles"
+          DependsOnTargets="_GetOutputItemsFromPack"
```

Comment:
@sharwell this is not as straightforward as you think:
1) Pack does not run as part of build, rather it runs after build, so you can't really make use of the i-built clean infrastrcture the way you are suggesting : https://github.com/Microsoft/msbuild/issues/2408#issuecomment-321082997 . The wiring up of FileWrites and writing the list to  *FileList.Absolute.txt happens at the end of build and before pack.

We could of course write to the *FileList.Absolute.txt ourselves, but that comes with it's own problems. https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L4898-L4906
Based on the above, the output nupkg is not under $(OutDir) and the output nuspec is in the BaseIntermediateOutputPath rather than IntermediateOutputPath.

3) We could write our own state file and do the clean up ourselves, but consider this scenario - if you are continuously building and packing with versions incrementing each time without running clean, the clean target would delete all the previously generated packages. Some people might want this to be the expected behavior, and some may not - we will wait to hear from community on this one , and if there is enough feedback, we will start planning on implementing it accordingly.